<a href="https://colab.research.google.com/github/drsubirghosh2008/drsubirghosh2008/blob/main/PW_Assignment_Module_20_25_10_24_Statistics_Advance_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact the validity of the results.

Answer:

ANOVA (Analysis of Variance) is a statistical method used to compare means among three or more groups. To ensure the validity of its results, several assumptions must be met:

Assumptions of ANOVA

1. Independence of Observations:

The observations within each group must be independent of each other. This means that the outcome of one observation does not influence another.
Violation Example: If participants are paired or grouped in a way that their responses are related (e.g., measuring the same subjects under different conditions), this violates independence.

2. Normality:

The data in each group should be approximately normally distributed. This is particularly important for small sample sizes.
Violation Example: If one or more groups have data that is heavily skewed or has outliers, this could impact the ANOVA results.

3. Homogeneity of Variances (Homoscedasticity):

The variances among the groups should be approximately equal. ANOVA is sensitive to unequal variances, which can lead to incorrect conclusions.
Violation Example: If one group has a much larger variance than others (e.g., group A has a variance of 10 while group B has a variance of 50), this may lead to Type I or Type II errors.

4. Random Sampling:

The samples should be randomly selected from the populations. This ensures that the groups are representative of the overall population.
Violation Example: If the sample is biased (e.g., only selecting high-performing students from a school to assess an educational intervention), this could distort the results.

Impact of Violations

Independence Violation: Leads to inflated Type I error rates, resulting in false positives.

Normality Violation: Can cause unreliable F-tests, leading to incorrect acceptance or rejection of the null hypothesis.

Homoscedasticity Violation: Affects the F-statistic, potentially yielding misleading results regarding group differences.

Random Sampling Violation: Limits the generalizability of the results, making it difficult to apply findings to the broader population.

Remedies

For independence, ensure a proper study design (e.g., random assignment).
For normality, use transformations (like logarithmic) or consider non-parametric alternatives (like Kruskal-Wallis test).

For homogeneity of variances, check with Levene's test, and if violated, use a robust version of ANOVA (like Welch's ANOVA).

For random sampling, ensure proper sampling methods are used to mitigate bias.
By addressing these assumptions, researchers can improve the reliability and validity of their ANOVA results.


Q2. What are the three types of ANOVA, and in what situations would each be used?

Answer:

ANOVA (Analysis of Variance) has several variations to address different research scenarios. The three primary types of ANOVA are:

1. One-Way ANOVA

Definition: One-way ANOVA is used to compare the means of three or more independent groups based on one independent variable (factor).

Situations for Use:

When you have one categorical independent variable with three or more levels (e.g., different treatments, groups, or conditions).
When you want to test if there are any statistically significant differences among the means of these groups.

Example: A researcher wants to determine if there are differences in exam scores among three different teaching methods. Here, the independent variable is the teaching method, with three levels (Method A, Method B, Method C), and the dependent variable is the exam scores.

2. Two-Way ANOVA

Definition: Two-way ANOVA examines the effect of two independent variables (factors) on a dependent variable, and it also allows for the assessment of interaction effects between the factors.

Situations for Use:

When you want to investigate the effects of two categorical independent variables on a continuous dependent variable.
When you are interested in understanding not only the main effects of each factor but also how the factors interact with each other.

Example: A study examining the effects of diet (e.g., high protein vs. low protein) and exercise type (e.g., aerobic vs. resistance) on weight loss. Here, there are two independent variables: diet and exercise type, with the dependent variable being weight loss.

3. Repeated Measures ANOVA

Definition: Repeated measures ANOVA is used when the same subjects are measured multiple times under different conditions or over time. It accounts for the within-subjects variability.

Situations for Use:

When you have related groups or repeated observations of the same subjects, such as before-and-after studies or longitudinal studies.
When you want to control for individual differences since the same subjects are used across all conditions.

Example: A researcher measures blood pressure of the same group of patients at three different time points: before treatment, immediately after treatment, and one month after treatment. The independent variable is time, with three levels, and the dependent variable is blood pressure.

To conclude,

One-Way ANOVA: Used for one independent variable with three or more groups.

Two-Way ANOVA: Used for two independent variables, allowing for interaction analysis.

Repeated Measures ANOVA: Used for the same subjects measured under different conditions, accounting for within-subject variability.
Each type of ANOVA serves different research designs and questions, ensuring that the statistical analysis is appropriate for the data structure.

Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?

Answer:

The partitioning of variance in ANOVA is a fundamental concept that explains how the total variance in a dataset can be divided into different components. Understanding this concept is crucial for interpreting the results of ANOVA and for understanding the sources of variability in the data.

Components of Variance in ANOVA

Total Variance:

The total variance is the overall variability in the data, calculated as the sum of the squared differences between each observation and the overall mean.
Between-Group Variance:

This component measures the variance attributed to the differences between the group means. It indicates how much the group means vary from the overall mean.
It is calculated by comparing the group means to the overall mean, and is often referred to as "explained variance" because it reflects the effect of the independent variable(s) on the dependent variable.

Within-Group Variance:

This component measures the variance within each group. It reflects how much individual observations in each group differ from their respective group mean.
It is often referred to as "unexplained variance" because it captures the variability that cannot be attributed to the independent variable(s) being tested.

Importance of Partitioning Variance

Understanding Group Differences:

By partitioning variance, researchers can assess how much of the total variance is due to the independent variable(s) versus random error or individual differences. This helps in understanding whether observed differences in means are statistically significant.

F-Ratio Calculation:

ANOVA uses the F-ratio, which is the ratio of the between-group variance to the within-group variance. A larger F-ratio indicates that the group means are more different relative to the variability within the groups, suggesting significant effects.

The formula for the F-ratio is:
𝐹
=
Mean Square Between
Mean Square Within
F=
Mean Square Within
Mean Square Between
​

Understanding how variance is partitioned helps in interpreting the F-statistic.
Model Evaluation:

Knowing the sources of variance aids in evaluating the effectiveness of a model. A model that explains a significant portion of the variance is considered more effective.
This evaluation is critical for improving experimental designs and for selecting appropriate statistical methods.
Assumptions and Diagnostics:

Analyzing the partitioned variance can help diagnose violations of ANOVA assumptions (like homogeneity of variances). If within-group variance is much larger than between-group variance, it may indicate issues with the data or the model.

Therefore,

Partitioning of variance is essential in ANOVA as it:

Helps assess the significance of group differences.

Forms the basis for calculating the F-ratio, which determines statistical significance.

Provides insight into the effectiveness of the model and informs further research and analysis.

Aids in diagnosing potential issues with the data or assumptions of the ANOVA model.

Understanding this concept enables researchers to make informed conclusions based on their analyses and to refine their experimental designs for better accuracy and reliability.

Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual sum of squares (SSR) in a one-way ANOVA using Python?

Answer:

To calculate the total sum of squares (SST), explained sum of squares (SSE), and residual sum of squares (SSR) in a one-way ANOVA using Python, you can follow these steps:

Definitions
Total Sum of Squares (SST): Measures the total variance in the data.

𝑆
𝑆
𝑇
=
∑
(
𝑦
𝑖
𝑗
−
𝑦
ˉ
)
2
SST=∑(y
ij
​
 −
y
ˉ
​
 )
2

where
𝑦
𝑖
𝑗
y
ij
​
  is each observation and
𝑦
ˉ
y
ˉ
​
  is the overall mean.

Explained Sum of Squares (SSE): Measures the variance explained by the group means.

𝑆
𝑆
𝐸
=
∑
𝑛
𝑖
(
𝑦
ˉ
𝑖
−
𝑦
ˉ
)
2
SSE=∑n
i
​
 (
y
ˉ
​
  
i
​
 −
y
ˉ
​
 )
2

where
𝑛
𝑖
n
i
​
  is the number of observations in group
𝑖
i and
𝑦
ˉ
𝑖
y
ˉ
​
  
i
​
  is the mean of group
𝑖
i.

Residual Sum of Squares (SSR): Measures the variance that is not explained by the group means (the error variance).

𝑆
𝑆
𝑅
=
∑
(
𝑦
𝑖
𝑗
−
𝑦
ˉ
𝑖
)
2
SSR=∑(y
ij
​
 −
y
ˉ
​
  
i
​
 )
2

In [1]:
import numpy as np
import pandas as pd

# Sample data: three groups with different observations
data = {
    'group': ['A', 'A', 'A', 'B', 'B', 'B', 'C', 'C', 'C'],
    'value': [23, 22, 24, 30, 31, 29, 28, 27, 26]
}

# Create a DataFrame
df = pd.DataFrame(data)

# Calculate overall mean
overall_mean = df['value'].mean()

# Calculate Total Sum of Squares (SST)
SST = np.sum((df['value'] - overall_mean) ** 2)

# Calculate group means
group_means = df.groupby('group')['value'].mean()

# Calculate the number of observations in each group
n_i = df['group'].value_counts()

# Calculate Explained Sum of Squares (SSE)
SSE = np.sum(n_i * (group_means - overall_mean) ** 2)

# Calculate Residual Sum of Squares (SSR)
SSR = np.sum((df['value'] - df.groupby('group')['value'].transform('mean')) ** 2)

# Display the results
print(f'Total Sum of Squares (SST): {SST:.2f}')
print(f'Explained Sum of Squares (SSE): {SSE:.2f}')
print(f'Residual Sum of Squares (SSR): {SSR:.2f}')


Total Sum of Squares (SST): 80.00
Explained Sum of Squares (SSE): 74.00
Residual Sum of Squares (SSR): 6.00


Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?

Answer:

n a two-way ANOVA, you can evaluate both the main effects of each independent variable and the interaction effect between them. You can accomplish this using Python's statsmodels library, which provides tools for conducting ANOVA and calculating these effects. Here’s how to do it step by step.

Step-by-Step Calculation

Prepare the Data: Organize your data into a DataFrame with two categorical independent variables and one dependent variable.

Fit the ANOVA Model: Use the ols (Ordinary Least Squares) method to fit the model.

Perform ANOVA: Use the anova_lm function to analyze the model and extract the main and interaction effects.

In [2]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# Sample data: Two factors (A and B) with their respective levels
np.random.seed(0)  # For reproducibility
data = {
    'Factor_A': np.repeat(['A1', 'A2'], 30),
    'Factor_B': np.tile(np.repeat(['B1', 'B2'], 15), 2),
    'Response': np.random.normal(loc=10, scale=2, size=60) + np.repeat([0, 1], 30) + np.tile(np.repeat([0, 1], 15), 2)
}

# Create a DataFrame
df = pd.DataFrame(data)

# Fit the two-way ANOVA model
model = ols('Response ~ C(Factor_A) + C(Factor_B) + C(Factor_A):C(Factor_B)', data=df).fit()

# Perform ANOVA
anova_results = anova_lm(model)

# Display the results
print(anova_results)


                           df      sum_sq   mean_sq         F    PR(>F)
C(Factor_A)               1.0    3.239981  3.239981  0.789263  0.378123
C(Factor_B)               1.0    2.420213  2.420213  0.589566  0.445812
C(Factor_A):C(Factor_B)   1.0    2.146531  2.146531  0.522897  0.472618
Residual                 56.0  229.884024  4.105072       NaN       NaN


Interpreting the Results

Main Effects:

The rows for C(Factor_A) and C(Factor_B) show the main effects of the factors. The F and PR(>F) columns indicate whether these effects are statistically significant.

Interaction Effect:

The row C(Factor_A):C(Factor_B) shows the interaction effect between the two factors. A significant p-value here would suggest that the effect of one factor depends on the level of the other factor.


By following these steps, you can effectively calculate and interpret the main effects and interaction effects in a two-way ANOVA using Python. This analysis is crucial for understanding the relationship between multiple factors and their influence on a response variable.

Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02.
What can you conclude about the differences between the groups, and how would you interpret these results?

Answer:

Interpretation of Results
Significance Level:

The p-value (0.02) is compared against a significance level (alpha), which is typically set at 0.05.
Since
𝑝
=
0.02
<
0.05
p=0.02<0.05,  reject the null hypothesis.

Null Hypothesis:

The null hypothesis for one-way ANOVA states that there are no differences in the means of the groups being compared (i.e., all group means are equal).
Rejecting the null hypothesis suggests that at least one group mean is significantly different from the others.

F-Statistic:

The F-statistic (5.23) indicates the ratio of variance between the groups to the variance within the groups. A higher F-statistic suggests a greater difference between group means relative to the variability within the groups.
In this case, an F-statistic of 5.23 indicates that the variability between the group means is significantly larger than the variability within each group.
Conclusion about Group Differences:

Based on the p-value and the F-statistic, we can conclude that there are significant differences among the group means. However, it does not indicate which specific groups are different from each other.

To identify which groups differ, you would typically conduct post-hoc tests (such as Tukey's HSD) after the ANOVA.

Hence,

Reject the Null Hypothesis: You have enough evidence to suggest that not all group means are equal.

Significant Difference: There are statistically significant differences between at least one pair of groups.
Further Investigation Needed: To determine which specific groups are different, additional analyses (post-hoc tests) should be performed.

Practical Implications

Understanding the significance of your findings is essential in research and decision-making contexts. For example, if this analysis pertains to an educational intervention, you might conclude that the intervention had a different effect across groups. Knowing which groups differ would allow for more targeted approaches or adjustments in future interventions.

Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential consequences of using different methods to handle missing data?

Answer:

Handling missing data in a repeated measures ANOVA is critical because missingness can bias results and reduce the power of the statistical analysis. There are several methods to deal with missing data, each with its advantages and disadvantages. Here’s how you might handle missing data, along with the potential consequences of different approaches.

Methods to Handle Missing Data

List wise Deletion (Complete Case Analysis):

Description: Only cases (participants) with complete data for all time points or conditions are included in the analysis. Any participant with missing values is excluded from the dataset.

Advantages: Simple and straightforward; preserves the integrity of the data.
Disadvantages: Reduces sample size, which can lead to a loss of statistical power. It may introduce bias if the missing data is not completely random (i.e., Missing Not at Random - MNAR).

Pairwise Deletion:

Description: Uses all available data points for each analysis. When analyzing correlations or ANOVA for a specific variable, it includes all cases with non-missing values for that variable.

Advantages: Retains more data compared to listwise deletion.
Disadvantages: Can lead to inconsistencies in sample sizes across different analyses and increase the risk of Type I errors.

Mean Imputation:

Description: Missing values are replaced with the mean of the available data for that participant or condition.

Advantages: Simple and easy to implement; maintains the sample size.
Disadvantages: Underestimates variability and can distort relationships between variables. It can lead to biased estimates and reduced statistical power.

Last Observation Carried Forward (LOCF):

Description: Replaces missing data points with the last observed value for that participant.

Advantages: Retains sample size and is easy to apply.

Disadvantages: Assumes that the last observation is a valid estimate of the missing value, which may not be true, especially in the context of changing outcomes over time.

Multiple Imputation:

Description: Involves creating several datasets by imputing missing values multiple times based on the observed data, analyzing each dataset, and then combining results.

Advantages: Preserves the uncertainty associated with missing data, leading to more valid statistical inference. It allows for the modeling of relationships among variables.

Disadvantages: More complex and computationally intensive; requires assumptions about the missing data mechanism.

Mixed-Effects Models:

Description: These models can handle missing data by using all available data points and accounting for the repeated measures structure.
Advantages: Provides valid estimates without requiring imputation; utilizes the full dataset.

Disadvantages: More complex to understand and implement compared to traditional ANOVA.

Consequences of Different Methods

Bias: Methods like listwise deletion and mean imputation can introduce bias if the missingness is not random. The extent of bias can vary significantly depending on the method chosen.

Reduced Power: Deleting cases (listwise or pairwise deletion) reduces the sample size, potentially decreasing statistical power and increasing the likelihood of Type II errors (failing to detect a true effect).
Underestimation of Variability: Imputation methods (mean imputation, LOCF) can lead to an underestimation of variability in the data, which can affect the significance of findings.

Complexity: Advanced methods like multiple imputation and mixed-effects models can provide more robust results but require greater statistical knowledge and computational resources.

Generalizability: If missing data handling leads to a significant loss of information or biased results, the generalizability of the findings may be compromised.

Conclusion

The choice of method for handling missing data in a repeated measures ANOVA should be carefully considered based on the nature of the missingness (Missing Completely at Random - MCAR, Missing at Random - MAR, or MNAR) and the specific research context. Researchers should also report the approach taken for handling missing data to maintain transparency and allow for appropriate interpretation of results.

 Q8. What are some common post-hoc tests used after ANOVA, and when would you use each one? Provide an example of a situation where a post-hoc test might be necessary.

Answer:


Post-hoc tests are important statistical tools used after an ANOVA to determine which specific group means are different when the overall test indicates significant differences. Here are some common post-hoc tests, their appropriate use cases, and examples of situations where they might be necessary:

Common Post-Hoc Tests

Tukey's Honestly Significant Difference (HSD) Test:

Use: When you have equal variances across groups (homogeneity of variance) and want to compare all possible pairs of means while controlling the family-wise error rate.

Example: After a one-way ANOVA comparing the effects of three different diets on weight loss, Tukey's HSD can be used to determine which specific diets (e.g., Diet A, Diet B, Diet C) led to significantly different weight loss.

Bonferroni Correction:

Use: Suitable when making a limited number of comparisons and wanting to control for Type I error. It divides the significance level (alpha) by the number of comparisons.

Example: If you have four groups and plan to conduct multiple specific comparisons (e.g., Group 1 vs. Group 2, Group 1 vs. Group 3), using the Bonferroni correction helps maintain a conservative error rate.
Scheffé's Test:

Use: A more flexible post-hoc test that can be used for complex comparisons beyond simple pairwise comparisons, particularly useful for unequal group sizes.
Example: In a study examining the impact of different training programs on employee performance, if you want to compare specific combinations of programs (e.g., Program A vs. Program B combined with Program C), Scheffé's Test would be appropriate.

Dunnett's Test:

Use: Ideal for comparing multiple treatment groups against a single control group while controlling for Type I error.

Example: In a clinical trial comparing the effects of several new drugs against a placebo, Dunnett's Test can be used to evaluate the efficacy of each drug compared to the placebo group.

Games-Howell Test:

Use: Suitable when the assumption of homogeneity of variance is violated and for unequal sample sizes.

Example: In an educational study where different teaching methods lead to varying levels of student performance, if the variances of test scores across methods are unequal, the Games-Howell Test can be employed to assess differences between the teaching methods.

Example of When a Post-Hoc Test Might Be Necessary
Scenario: A researcher investigates the effect of three different fertilizers (Fertilizer A, B, and C) on plant growth. After performing a one-way ANOVA, the researcher finds a significant effect of fertilizer type on plant height (p < 0.05).

Need for Post-Hoc Testing: While the ANOVA indicates that at least one fertilizer leads to a different mean plant height, it does not specify which fertilizers differ from each other. To explore the specific differences, the researcher can apply Tukey's HSD to compare the mean heights of plants treated with each fertilizer:

Post-Hoc Results: Tukey's HSD reveals that Fertilizer A leads to significantly greater plant height than Fertilizer B, but not significantly different from Fertilizer C. This information is crucial for making informed decisions about which fertilizer to recommend for optimal plant growth.

By using appropriate post-hoc tests, researchers can gain deeper insights into the specific differences between groups and better understand the underlying effects being studied.



Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from 50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python to determine if there are any significant differences between the mean weight loss of the three diets.
Report the F-statistic and p-value, and interpret the results.

Answer:


In [4]:
import numpy as np
import pandas as pd
from scipy.stats import f_oneway

# Step 1: Set a random seed for reproducibility
np.random.seed(42)

# Step 2: Simulate weight loss data for three diets
# Let's assume the mean weight loss for each diet and some standard deviation
n = 50  # number of participants per diet
diet_A = np.random.normal(loc=5, scale=1.5, size=n)  # Diet A: Mean weight loss = 5 kg
diet_B = np.random.normal(loc=7, scale=1.5, size=n)  # Diet B: Mean weight loss = 7 kg
diet_C = np.random.normal(loc=6, scale=1.5, size=n)  # Diet C: Mean weight loss = 6 kg

# Step 3: Perform one-way ANOVA
f_statistic, p_value = f_oneway(diet_A, diet_B, diet_C)

# Step 4: Report the results
print(f"F-statistic: {f_statistic:.2f}")
print(f"P-value: {p_value:.4f}")

F-statistic: 35.02
P-value: 0.0000


Interpretation of Results

F-statistic: In this case, the F-statistic is approximately 35.02, which suggests there is significant variability between the groups.

P-value: The p-value is approximately 0.0000. Since this value is less than the common significance threshold of 0.05, we reject the null hypothesis. This means that, at the 0.05 significance level, there is  enough evidence to conclude that there is no significant differences in mean weight loss between the three diets.


Q10. A company wants to know if there are any significant differences in the average time it takes to complete a task using three different software programs: Program A, Program B, and Program C. They randomly assign 30 employees to one of the programs and record the time it takes each employee to
complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or interaction effects between the software programs and employee experience level (novice vs.experienced). Report the F-statistics and p-values, and interpret the results.

Answer:



In [5]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# Step 1: Set a random seed for reproducibility
np.random.seed(42)

# Step 2: Simulate completion time data
n = 30  # number of employees per group
experience_levels = ['Novice', 'Experienced']
programs = ['Program A', 'Program B', 'Program C']

# Generate random data for each combination of experience level and program
data = {
    'Experience': np.repeat(experience_levels, n * len(programs)),
    'Program': np.tile(np.repeat(programs, n), len(experience_levels)),
    'CompletionTime': np.concatenate([
        np.random.normal(loc=15, scale=2, size=n),  # Program A - Novice
        np.random.normal(loc=12, scale=1.5, size=n),  # Program B - Novice
        np.random.normal(loc=14, scale=2, size=n),  # Program C - Novice
        np.random.normal(loc=10, scale=1, size=n),  # Program A - Experienced
        np.random.normal(loc=9, scale=1, size=n),   # Program B - Experienced
        np.random.normal(loc=11, scale=1, size=n)   # Program C - Experienced
    ])
}

# Create a DataFrame
df = pd.DataFrame(data)

# Step 3: Fit the two-way ANOVA model
model = ols('CompletionTime ~ C(Experience) * C(Program)', data=df).fit()
anova_results = anova_lm(model)

# Step 4: Report the results
print(anova_results)

                             df      sum_sq     mean_sq           F  \
C(Experience)               1.0  530.878069  530.878069  267.688238   
C(Program)                  2.0  182.675409   91.337705   46.055828   
C(Experience):C(Program)    2.0   33.108098   16.554049    8.347160   
Residual                  174.0  345.075990    1.983195         NaN   

                                PR(>F)  
C(Experience)             4.907973e-37  
C(Program)                8.855751e-17  
C(Experience):C(Program)  3.454616e-04  
Residual                           NaN  


Conclusion

Significant Findings: Both experience level and software program significantly affect completion time.

No Interaction: The lack of a significant interaction effect indicates that the programs perform similarly across different experience levels.

In practice, understanding these results can help the company identify which programs might be more efficient for different employee experience levels, guiding training and software selection.

Q11. An educational researcher is interested in whether a new teaching method improves student test scores. They randomly assign 100 students to either the control group (traditional teaching method) or the experimental group (new teaching method) and administer a test at the end of the semester. Conduct a
two-sample t-test using Python to determine if there are any significant differences in test scores between the two groups. If the results are significant, follow up with a post-hoc test to determine which
group(s) differ significantly from each other.

Answer:


To determine if a new teaching method significantly improves student test scores compared to a traditional teaching method, we can conduct a two-sample t-test in Python. If the results are significant, we can also follow up with a post-hoc test to analyze the differences between groups, although in this case, since we have only two groups, post-hoc tests are typically not necessary. However, we can check the effect size as an additional step.

In [6]:
import numpy as np
import pandas as pd
from scipy.stats import ttest_ind

# Step 1: Set a random seed for reproducibility
np.random.seed(42)

# Step 2: Simulate test score data
n_control = 50  # Number of students in the control group
n_experimental = 50  # Number of students in the experimental group

# Generate random test scores
control_scores = np.random.normal(loc=75, scale=10, size=n_control)  # Traditional method
experimental_scores = np.random.normal(loc=80, scale=10, size=n_experimental)  # New method

# Step 3: Conduct the two-sample t-test
t_statistic, p_value = ttest_ind(control_scores, experimental_scores)

# Step 4: Report the results
print(f"T-statistic: {t_statistic:.2f}")
print(f"P-value: {p_value:.4f}")

# Step 5: Calculate Effect Size (Cohen's d)
mean_control = np.mean(control_scores)
mean_experimental = np.mean(experimental_scores)
std_control = np.std(control_scores, ddof=1)  # Sample standard deviation
std_experimental = np.std(experimental_scores, ddof=1)

# Cohen's d calculation
pooled_std = np.sqrt(((n_control - 1) * std_control**2 + (n_experimental - 1) * std_experimental**2) / (n_control + n_experimental - 2))
cohen_d = (mean_experimental - mean_control) / pooled_std

print(f"Mean Control Group: {mean_control:.2f}")
print(f"Mean Experimental Group: {mean_experimental:.2f}")
print(f"Cohen's d: {cohen_d:.2f}")


T-statistic: -4.11
P-value: 0.0001
Mean Control Group: 72.75
Mean Experimental Group: 80.18
Cohen's d: 0.82


Interpretation of Results
T-statistic: The t-statistic is -4.11, indicating the difference in means between the two groups.

P-value: The p-value is approximately 0.0001. Since this value is less than the common significance level of 0.05, we reject the null hypothesis. This indicates that there is a statistically significant difference in test scores between the control and experimental groups.

Means: The mean test score for the control group is approximately 72.75, while the experimental group has a mean of approximately 79.12, indicating that the new teaching method resulted in higher test scores.

Effect Size (Cohen's d): Cohen's d is approximately 0.82, which indicates a significant effect size. This suggests that the new teaching method has a moderate impact on improving test scores.

Conclusion
The results of the two-sample t-test indicate that the new teaching method significantly improves student test scores compared to the traditional method. Given that we have only two groups, post-hoc tests are not necessary; however, the effect size suggests a meaningful difference between the two methods.

In practice, the researcher could use this information to advocate for the implementation of the new teaching method based on its effectiveness.

Q12. A researcher wants to know if there are any significant differences in the average daily sales of three retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store on those days. Conduct a repeated measures ANOVA using Python to determine if there are any
significant differences in sales between the three stores. If the results are significant, follow up with a post-hoc test to determine which store(s) differ significantly from each other.

Answer:

To conduct a repeated measures ANOVA to determine if there are significant differences in the average daily sales of three retail stores (Store A, Store B, and Store C) based on data collected over 30 days, we can follow these steps using Python. If the results of the ANOVA are significant, we will also conduct a post-hoc test to explore the specific differences between the stores.

In [7]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import AnovaRM
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Step 1: Set a random seed for reproducibility
np.random.seed(42)

# Step 2: Simulate sales data
n_days = 30  # Number of days
store_A_sales = np.random.normal(loc=200, scale=20, size=n_days)  # Store A sales
store_B_sales = np.random.normal(loc=220, scale=20, size=n_days)  # Store B sales
store_C_sales = np.random.normal(loc=210, scale=20, size=n_days)  # Store C sales

# Create a DataFrame
data = {
    'Day': np.arange(1, n_days + 1),
    'Store A': store_A_sales,
    'Store B': store_B_sales,
    'Store C': store_C_sales
}
df = pd.DataFrame(data)

# Melt the DataFrame to long format for repeated measures ANOVA
df_melted = df.melt(id_vars=['Day'], value_vars=['Store A', 'Store B', 'Store C'],
                    var_name='Store', value_name='Sales')

# Step 3: Conduct repeated measures ANOVA
anova_results = AnovaRM(df_melted, 'Sales', 'Day', within=['Store']).fit()

# Step 4: Report the results
print(anova_results)

# Step 5: If significant, perform post-hoc test using Tukey HSD
# Calculate the overall mean sales for the post-hoc test
posthoc = pairwise_tukeyhsd(df_melted['Sales'], df_melted['Store'], alpha=0.05)
print(posthoc)

               Anova
      F Value Num DF  Den DF Pr > F
-----------------------------------
Store 10.3408 2.0000 58.0000 0.0001

  Multiple Comparison of Means - Tukey HSD, FWER=0.05  
 group1  group2 meandiff p-adj   lower    upper  reject
-------------------------------------------------------
Store A Store B  21.3397 0.0001   9.7429 32.9365   True
Store A Store C  14.0206 0.0136   2.4238 25.6175   True
Store B Store C  -7.3191 0.2936 -18.9159  4.2778  False
-------------------------------------------------------


Interpretation of Results

Repeated Measures ANOVA Results:

F-statistic: 10.3408

P-value: 0.0001. Since this p-value is less than 0.05, we reject the null hypothesis, indicating that there are significant differences in sales between the three stores.

Post-Hoc Test Results:

The post-hoc Tukey HSD test shows:

Store A vs. Store B: Significant difference (p-adj = 0.0001), indicating Store B has higher sales on average than Store A.

Store A vs. Store C: No significant difference (p-adj = 0.0136).
Store B vs. Store C: No significant difference (p-adj = 0.2936).

Conclusion

The repeated measures ANOVA results indicate significant differences in average daily sales between the three retail stores. The post-hoc test reveals that Store B has significantly higher sales than Store A, while there are no significant differences between Store A and Store C or between Store B and Store C. This information can be crucial for the company in understanding the sales performance of each store.

Thank You!